### PR TITLE
Proof function util helpers refactor

### PR DESCRIPTION
**Pull request by scorbajio**
This change is an alternative approach to the work started in #3538. It looks into refactoring the trie class to remove proof functionality from the main class and move them into standalone functions in order to improve tree shaking for the trie package and decrease bundle size.

### CODE REVIEW

**Comment by scorbajio** [comment]
Using a minimal code path for the `Trie` class:
```
import { Trie } from '@ethereumjs/trie'

const t = new Trie()
console.log(t.root())
```
Results:
```
before: 223.6kb
after: 205.7kb
delta: -17.9kb
```

File: packages/trie/src/proof/index.ts [hunk 1]
```diff
@@ -64,4 +66,124 @@ export function verifyTrieRangeProof(
...
+}
+
+/**
+ * Create a trie from a given (EIP-1186)[https://eips.ethereum.org/EIPS/eip-1186] proof. An EIP-1186 proof contains the encoded trie nodes from the root
+ * node to the leaf node storing state data. This function does not check if the proof has the same expected root. A static version of this function exists
+ * with the same name.
+ * @param proof an EIP-1186 proof to update the trie from
+ * @deprecated Use `updateFromProof`
+ */
+export async function fromProof(trie: Trie, proof: Proof): Promise<void> {
```
Comment (acolytec3) [inline packages/trie/src/proof/index.ts:176]: This function feels duplicative of `createTrieFromProof` in the `constructors.ts`.  Can you think of a reason we'd need it here as well?

File: packages/trie/src/proof/index.ts [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
 import { createTrieFromProof } from '../constructors.js'
-import { verifyRangeProof } from '../index.js'
+import { Trie, verifyRangeProof } from '../index.js'
 import { bytesToNibbles } from '../util/nibbles.js'
 
 import type { Proof, TrieOpts } from '../index.js'
+import type { PutBatch } from '@ethereumjs/util'
 
 /**
  * Static version of verifyProof function with the same behavior. An (EIP-1186)[https://eips.ethereum.org/EIPS/eip-1186] proof contains the encoded trie nodes
```
Comment (acolytec3) [inline packages/trie/src/proof/index.ts:12]: This phrasing is obsolete
```suggestion
 * An (EIP-1186)[https://eips.ethereum.org/EIPS/eip-1186] proof contains the encoded trie nodes
```

File: packages/trie/src/proof/index.ts [hunk 3]
```diff
@@ -64,4 +66,124 @@ export function verifyTrieRangeProof(
...
+
+  await trie['_db'].batch(opStack)
+  if (opStack[0] !== undefined) {
+    return opStack[0].key
+  }
+}
+
+/**
+ * Verifies a proof by putting all of its nodes into a trie and attempting to get the proven key. An (EIP-1186)[https://eips.ethereum.org/EIPS/eip-1186] proof
+ * contains the encoded trie nodes from the root node to the leaf node storing state data. A static version of this function exists with the same name.
```
Comment (acolytec3) [inline packages/trie/src/proof/index.ts:121]:
```suggestion
 * contains the encoded trie nodes from the root node to the leaf node storing state data.
```

File: packages/trie/test/util/log.spec.ts [hunk 4]
```diff
@@ -34,7 +35,7 @@ describe('Run Trie script with DEBUG enabled', async () => {
   process.env.DEBUG = ''
   const trie2 = new Trie({})
   trie2['DEBUG'] = true
-  await trie2.fromProof(proof)
+  await fromProof(trie2, proof)
```
Comment (acolytec3) [inline packages/trie/test/util/log.spec.ts:38]:
```suggestion
  await createTrieFromProof(trie2, proof)
```

File: packages/trie/test/util/log.spec.ts [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import { utf8ToBytes } from '@ethereumjs/util'
 import { assert, describe, it } from 'vitest'
 
+import { createProof, fromProof, verifyProof } from '../../src/index.js'
```
Comment (acolytec3) [inline packages/trie/test/util/log.spec.ts:4]:
```suggestion
import { createProof, createTrieFromProof, verifyProof } from '../../src/index.js'
```

**Review by acolytec3 — CHANGES_REQUESTED** [review]
Looks great overall.  Some comments need cleanup and I think we can remove that `fromProof` function entirely.

File: packages/trie/src/proof/index.ts (hunk 1 above)
Comment (scorbajio) [reply packages/trie/src/proof/index.ts]: I'd agree

File: packages/trie/test/util/log.spec.ts (hunk 4 above)
Comment (scorbajio) [reply packages/trie/test/util/log.spec.ts]: Fixed locally

File: packages/trie/test/util/log.spec.ts (hunk 5 above)
Comment (scorbajio) [reply packages/trie/test/util/log.spec.ts]: Fixed locally

File: packages/trie/src/proof/index.ts [hunk 6]
```diff
@@ -64,4 +65,104 @@ export function verifyTrieRangeProof(
...
+  const p = stack.map((stackElem) => {
+    return stackElem.serialize()
+  })
+  trie['DEBUG'] && trie['debug'](`Proof created with (${stack.length}) nodes`, ['CREATE_PROOF'])
+  return p
+}
+
+/**
+ * Updates a trie from a proof by putting all the nodes in the proof into the trie. If a trie is being updated with multiple proofs, {@param shouldVerifyRoot} can
+ * be passed as false in order to not immediately throw on an unexpected root, so that root verification can happen after all proofs and their nodes have been added.
```
Comment (gabrocheleau) [inline packages/trie/src/proof/index.ts:86]: This comment seems outdated, mentioning that `{@param shouldVerifyRoot} can be passed as false` seems to imply that it does not default to false (which it does). Imo we should clarify this, and add a mention the the `shouldVerifyRoot` param natspec that it defaults to false (I would naturally assume that root is verified).

**Review by acolytec3 — APPROVED** [review]
LGTM

**Pull request merged**